### PR TITLE
Updated RegistrationController for extensibility

### PR DESCRIPTION
**Pull request by opichon**
RegistrationController updated for extensibility. Controller actions are expected to obtain the request via a function argument. The `getRequest` method will deprecated in a future version of symfony.

Some methods of RegistrationController don't need te request (`checkEmailAction`, `confirmedAction`). However, if the controller is extended via a sub-class, and the methods of that class do need the request, then there is no solution.

Adding the request as a function argument to the original class's methods, even if it is not used, resolves this and allows the controller class to be extended easily.

### CODE REVIEW

**Comment by core23** [comment]
Good catch

**Comment by opichon** [comment]
Also made `getTargetUrlFromSession` method protected. Since it is used in `confirmedAction`, a child class extending RegistrationController may want to invoke it.

File: Controller/RegistrationController.php [hunk 1]
```diff
@@ -96,7 +96,7 @@ public function registerAction(Request $request)
     /**
      * Tell the user to check his email provider.
      */
-    public function checkEmailAction()
+    public function checkEmailAction(Request $request)
```
Comment (dmaicher) [inline Controller/RegistrationController.php]: whenever people overwrite this method they could simply get the request from the `request_stack` though :wink:

Comment (opichon) [inline Controller/RegistrationController.php]: That is correct, but isn't the proposed change just more extensible?

Comment (dmaicher) [inline Controller/RegistrationController.php]: It's just that you wrote:

"However, if the controller is extended via a sub-class, and the methods of that class do need the request, **then there is no solution**."

Which is not correct :wink:

Comment (core23) [inline Controller/RegistrationController.php]: It should, but it doesn't need the same signature: https://3v4l.org/ULUpv 

cc @greg0ire

Comment (greg0ire) [inline Controller/RegistrationController.php]: Can someone check have symfony behaves when you add `Request $request = null` ? Is the `$request` filled? If yes, this should be closed (or at least shrinked).

Comment (opichon) [inline Controller/RegistrationController.php]: I get this error:

```
Warning: Declaration of UAM\Bundle\UserBundle\Controller\RegistrationController::checkEmailAction(Symfony\Component\HttpFoundation\Request $request) should be compatible with FOS\UserBundle\Controller\RegistrationController::checkEmailAction()
```

Comment (greg0ire) [inline Controller/RegistrationController.php]: I don't see `= null` in your error message.

Comment (opichon) [inline Controller/RegistrationController.php]: Apologies, I thought you meant something else. With `Request $request = null` the page seems to work. This is not an exhaustive test though.

However, I believe the recommended signature in symfony 3.0+ is `Request $request`. Why would the request be null? Are you concerned about backward compatibility with existing code? Why would that action be invoked with a null request?

Comment (greg0ire) [inline Controller/RegistrationController.php]: I explained it all [here](https://github.com/sonata-project/SonataNewsBundle/pull/342#issuecomment-255391982)
**TL;DR** : you don't need to add a $request parameter on every action for it to be "extensible"

File: Controller/RegistrationController.php [hunk 2]
```diff
@@ -178,7 +178,7 @@ public function confirmedAction()
     /**
      * @return mixed
      */
-    private function getTargetUrlFromSession()
+    protected function getTargetUrlFromSession()
```
Comment (greg0ire) [inline Controller/RegistrationController.php]: Only this change should be in the PR

Comment (stof) [inline Controller/RegistrationController.php]: I don't want to make this method an extension point. It is not meant to be one

File: Controller/RegistrationController.php [hunk 3]
```diff
@@ -162,7 +162,7 @@ public function confirmAction(Request $request, $token)
     /**
      * Tell the user his account is now confirmed.
      */
-    public function confirmedAction()
+    public function confirmedAction(Request $request)
```
Comment (stof) [inline Controller/RegistrationController.php]: this is a BC break for anyone extending the controller already, as it changes the signature.

And it does not provide you new features, as using the RequestStack already allows you to get the Request. So I'm -1 on this PR.

**Comment by opichon** [comment]
@stof I have a use case where I am overriding the `RegistrationController#confirmedAction`. That method uses `getTargetUrl`. If `confirmedAction` uses a private action, then it also is not meant to be an extension point. Does that make sense?

**Comment by greg0ire** [comment]
@opichon @stof : a solution to make it usable without becoming an extension point would be to make it `final protected`

**Comment by opichon** [comment]
@stof It's your call, of course.

I appreciate that adding the request to the RegistrationController actions does indeed cause a BC issue with anyone who has already extended the RegistrationController class. However, I can see that plenty of changes already in master are going to cause considerable BC issues. And the bundle is still in alpha. I find this weakens the BC argument somewhat.

In fact, this is why I strongly advise (and require from my developers) that the request should be passed systematically as the first argument of all controller action methods. This eliminates from the start all these issues. 

However, since I have a working solution, it's not a big deal for me now.

As to the getTargetUrlFromSession method, the issue is not so much to allow overriding the `getTargetUrlFromSession` method, for which I cannot at this point envision any legitimate need, as to make it available to subclasses. It seems a useful method.

**Comment by XWB** [comment]
I'm with @stof, -1 on this one.

**Comment by greg0ire** [comment]
> a solution to make it usable without becoming an extension point would be to make it final protected

@XWB @stof @opichon : thoughts?
